src/Common/src/TypeSystem/IL/TypeSystemContext.ValueTypeMethods.cs

```diff
@@ -0,0 +1,138 @@
+            protected override bool CompareKeyToValue(DefType key, MethodDesc value) => key == value.OwningType;
+            protected override bool CompareValueToValue(MethodDesc v1, MethodDesc v2) => v1.OwningType == v2.OwningType;
+
+            protected override MethodDesc CreateValueFromKey(DefType key)
+            {
+                return new ValueTypeGetFieldHelperMethodOverride(key);
+            }
+        }
+
+        private ValueTypeMethodHashtable _valueTypeMethodHashtable = new ValueTypeMethodHashtable();
+
+        protected virtual IEnumerable<MethodDesc> GetAllMethodsForValueType(TypeDesc valueType)
+        {
+            TypeDesc valueTypeDefinition = valueType.GetTypeDefinition();
+
+            if (RequiresGetFieldHelperMethod((MetadataType)valueTypeDefinition))
+            {
+                MethodDesc getFieldHelperMethod = _valueTypeMethodHashtable.GetOrCreateValue((DefType)valueTypeDefinition);
+
+                // Check that System.ValueType has the method we're overriding.
+                Debug.Assert(valueTypeDefinition.BaseType.GetMethod(getFieldHelperMethod.Name, null) != null);
+
+                if (valueType != valueTypeDefinition)
+                {
+                    yield return GetMethodForInstantiatedType(getFieldHelperMethod, (InstantiatedType)valueType);
+                }
+                else
+                {
+                    yield return getFieldHelperMethod;
+                }
+            }
+
+            foreach (MethodDesc method in valueType.GetMethods())
+                yield return method;
+        }
+
+        private bool RequiresGetFieldHelperMethod(MetadataType valueType)
+        {
+            if (_objectGetHashCodeMethod == null)
+                _objectGetHashCodeMethod = GetWellKnownType(WellKnownType.Object).GetMethod("GetHashCode", null);
+
+            if (_objectEqualsMethod == null)
+                _objectEqualsMethod = GetWellKnownType(WellKnownType.Object).GetMethod("Equals", null);
+
+            // If the classlib doesn't have Object.Equals/Object.GetHashCode, we don't need this.
+            if (_objectEqualsMethod == null && _objectGetHashCodeMethod == null)
+                return false;
+
+            // Byref-like valuetypes cannot be boxed.
+            if (valueType.IsByRefLike)
+                return false;
+
+            // Enums get their overrides from System.Enum.
+            if (valueType.IsEnum)
+                return false;
+
+            return !CanCompareValueTypeBits(valueType);
+        }
+
+        private bool CanCompareValueTypeBits(MetadataType type)
+        {
+            Debug.Assert(type.IsValueType);
+
+            if (type.ContainsGCPointers)
+                return false;
+
+            // TODO: what we're shooting for is overlapping fields
+            // or gaps between fields
+            if (type.IsExplicitLayout || type.GetClassLayout().Size != 0)
```

Comment:
What happens for padding from alignment? For example:
```
struct MyStruct
{
    byte b; // byte 0
    // 3-7 bytes of padding
    IntPtr ptr; // byte 4 or 8
}
```

We wouldn't want to do byte comparisons with padding since it's not guaranteed to be set

Reply:
Fixed the TODO.